workflow/executor/executor.go

```diff
@@ -930,150 +927,45 @@ func (we *WorkflowExecutor) Wait(ctx context.Context) error {
-	pod, err := we.getPod(ctx)
-	if err != nil {
-		log.Warnf("Failed to set execution control from API server: %v", err)
-		return
-	}
-	execCtlString, ok := pod.ObjectMeta.Annotations[common.AnnotationKeyExecutionControl]
-	if !ok {
-		we.ExecutionControl = nil
-	} else {
-		var execCtl common.ExecutionControl
-		err = json.Unmarshal([]byte(execCtlString), &execCtl)
-		if err != nil {
-			log.Errorf("Error unmarshalling '%s': %v", execCtlString, err)
-			return
-		}
-		we.ExecutionControl = &execCtl
-		log.Infof("Execution control set from API: %v", *we.ExecutionControl)
-	}
-}
-
 // monitorDeadline checks to see if we exceeded the deadline for the step and
 // terminates the main container if we did
-func (we *WorkflowExecutor) monitorDeadline(ctx context.Context, containerNames []string, annotationsUpdate <-chan struct{}) {
+func (we *WorkflowExecutor) monitorDeadline(ctx context.Context, containerNames []string) {
+	terminate := make(chan os.Signal, 1)
+	signal.Notify(terminate, syscall.SIGTERM)
+
+	deadlineExceeded := make(chan bool, 1)
+	if !we.Deadline.IsZero() {
+		t := time.AfterFunc(time.Until(we.Deadline), func() {
+			deadlineExceeded <- true
+		})
+		defer t.Stop()
+	}
+
 	log.Infof("Starting deadline monitor")
 	for {
 		select {
 		case <-ctx.Done():
 			log.Info("Deadline monitor stopped")
 			return
-		case <-annotationsUpdate:
-		default:
-			// TODO(jessesuen): we do not effectively use the annotations update channel yet. Ideally, we
-			// should optimize this logic so that we use some type of mutable timer against the deadline
-			// value instead of polling.
-			if we.ExecutionControl != nil && we.ExecutionControl.Deadline != nil {
-				if time.Now().UTC().After(*we.ExecutionControl.Deadline) {
-					var message string
-
-					// Zero value of the deadline indicates an intentional cancel vs. a timeout. We treat
-					// timeouts as a failure and the pod should be annotated with that error
-					if we.ExecutionControl.Deadline.IsZero() {
-						message = "terminated"
-					} else {
-						message = "Step exceeded its deadline"
-					}
-					log.Info(message)
-					util.WriteTeriminateMessage(message)
-					log.Infof("Killing main container")
-					terminationGracePeriodDuration, _ := we.GetTerminationGracePeriodDuration(ctx)
-					err := we.RuntimeExecutor.Kill(ctx, containerNames, terminationGracePeriodDuration)
-					if err != nil {
-						log.Warnf("Failed to kill main container: %v", err)
-					}
-					return
-				}
-			}
-			time.Sleep(1 * time.Second)
+		case <-deadlineExceeded:
+			we.killMainContainer(ctx, containerNames, "Step exceeded its deadline")
+			return
+		case <-terminate:
+			we.killMainContainer(ctx, containerNames, "Step terminated")
+			return
 		}
 	}
 }
 
+func (we *WorkflowExecutor) killMainContainer(ctx context.Context, containerNames []string, message string) {
+	log.Infof("Killing main container: %s", message)
+	util.WriteTeriminateMessage(message)
```

Comment:
huh - this line seem incongruous to me - setting the messages should probably be in another func

Reply:
Moved the messages back to the `monitorDeadline()` function, which does all the logging. Now `killContainers()` just does what it is supposed to do.